docker/Dockerfile

```diff
@@ -1,46 +1,84 @@
+ENV LANG=$LANG
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN \
+    apt-get update -q -y && \
+    apt-get upgrade -q -y && \
+    apt-get install -q --no-install-recommends -y \
+        build-essential \
+        ca-certificates \
+        libsasl2-dev \
+        locales \
+        openjdk-11-jdk && \
+    rm -rf /var/lib/apt/lists/* && \
+    locale-gen ${LANG}
+
+##############################
 FROM base AS dashboard-builder
 
-RUN apt-get install -y curl gnupg protobuf-compiler && mkdir -p /etc/apt/keyrings \
-  && curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg \
-  && echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list \
-  && apt-get update && apt-get install -y nodejs
-
 COPY ./dashboard/ /risingwave/dashboard
 COPY ./proto  /risingwave/proto
 
-RUN cd /risingwave/dashboard && npm i && npm run build-static && rm -rf node_modules
-
+WORKDIR /risingwave/dashboard
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN \
+    apt-get update -q -y && \
+    apt-get install -q --no-install-recommends -y \
+        curl \
+        gnupg \
+        libprotobuf-dev \
+        protobuf-compiler && \
+    mkdir -p /etc/apt/keyrings && \
+    curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg && \
+    echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list && \
+    apt-get update -q -y && \
+    apt-get install -q --no-install-recommends -y nodejs && \
+    rm -rf /var/lib/apt/lists/* && \
+    npm i && \
+    npm run build-static && \
+    rm -rf node_modules
+
+
+#######################
 FROM base AS rust-base
 
-RUN apt-get -y install make cmake protobuf-compiler curl bash lld unzip
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN \
+    apt-get update -q -y && \
+    apt-get install -q --no-install-recommends -y \
+        cmake \
+        curl \
+        libprotobuf-dev \
+        lld \
+        protobuf-compiler \
+        unzip && \
+    rm -rf /var/lib/apt/lists/* && \
+    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --no-modify-path --default-toolchain none -y && \
+    mkdir -p /risingwave
+
+WORKDIR /risingwave
 
-SHELL ["/bin/bash", "-c"]
+COPY rust-toolchain rust-toolchain
 
-RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --no-modify-path --default-toolchain none -y
 ENV PATH /root/.cargo/bin/:$PATH
 ENV CARGO_INCREMENTAL=0
 
-COPY rust-toolchain rust-toolchain
-
 # We need to add the `rustfmt` dependency, otherwise `risingwave_pb` will not compile
-RUN rustup self update \
-  && rustup set profile minimal \
-  && rustup show \
-  && rustup component add rustfmt
-
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN rustup set profile minimal && \
```

Comment:
I'm not sure whether `rustup self update` is needed, but IIRC `rustup show` is needed, which installs the toolchain. 🤔️

Reply:
I ran all these commands by hand when I was testing this before putting them into the Dockerfile.  `rustup self update` didn't do anything (nothing to update), and `rustup show` seemed to just print what was already there, also without doing anything.  Since it built ok and ran and the tests passed without running it, I assumed it was just a diagnositic for the build logs.  I can put that back if it is actually doing something, no problem.

Reply:
If you already installed rust toolchain, `rustup show` doesn't do anything. If you didn't install it, the command will install it for you. BTW, `cargo` commands will also install toolchain if you didn't.